core/src/Kokkos_HPX.hpp

```diff
@@ -215,120 +169,52 @@ class HPX {
-  HPX(const HPX &other)       = default;
-  HPX &operator=(const HPX &other) = default;
-#else
-  HPX() noexcept {}
-#endif
+      : m_instance_data(
+            mode == instance_mode::independent
+                ? (std::make_shared<instance_data>(m_next_instance_id++))
+                : std::shared_ptr<instance_data>(&m_default_instance_data,
+                                                 default_instance_deleter)) {}
+  HPX(hpx::execution::experimental::unique_any_sender<> &&sender)
+      : m_instance_data(std::make_shared<instance_data>(m_next_instance_id++,
+                                                        std::move(sender))) {}
 
-  void print_configuration(std::ostream &os, bool /*verbose*/ = false) const {
-    os << "HPX backend\n";
-    os << "HPX Execution Space:\n";
-    os << "  KOKKOS_ENABLE_HPX: yes\n";
-    os << "\nHPX Runtime Configuration:\n";
-  }
-  uint32_t impl_instance_id() const noexcept { return m_instance_id; }
+  HPX(HPX &&other)      = default;
+  HPX(const HPX &other) = default;
 
-#if defined(KOKKOS_ENABLE_HPX_ASYNC_DISPATCH)
-  static bool in_parallel(HPX const &instance = HPX()) noexcept {
-    return !instance.impl_get_future().is_ready();
-  }
-#else
-  static bool in_parallel(HPX const & = HPX()) noexcept { return false; }
-#endif
+  HPX &operator=(HPX &&) = default;
+  HPX &operator=(const HPX &) = default;
 
-#if defined(KOKKOS_ENABLE_HPX_ASYNC_DISPATCH)
-  static void impl_decrement_active_parallel_region_count() {
-    std::unique_lock<hpx::spinlock> l(m_active_parallel_region_count_mutex);
-    if (--m_active_parallel_region_count == 0) {
-      l.unlock();
-      m_active_parallel_region_count_cond.notify_all();
-    };
+  void print_configuration(std::ostream &os, bool /*verbose*/ = false) const;
+  instance_data &impl_get_instance_data() const noexcept {
+    KOKKOS_ASSERT(m_instance_data.get());
+    return *m_instance_data.get();
   }
-
-  static void impl_increment_active_parallel_region_count() {
-    std::unique_lock<hpx::spinlock> l(m_active_parallel_region_count_mutex);
-    ++m_active_parallel_region_count;
+  uint32_t impl_instance_id() const noexcept {
+    return impl_get_instance_data().m_instance_id;
   }
-#endif
 
-  void fence(
-      const std::string &name =
-          "Kokkos::Experimental::HPX::fence: Unnamed Instance Fence") const {
-    Kokkos::Tools::Experimental::Impl::profile_fence_event<
-        Kokkos::Experimental::HPX>(
-        name,
-        Kokkos::Tools::Experimental::Impl::DirectFenceIDHandle{
-            impl_instance_id()},
-        [&]() {
-#if defined(KOKKOS_ENABLE_HPX_ASYNC_DISPATCH)
-          impl_get_future().wait();
-          // Reset the future to free variables that may have been captured in
-          // parallel regions.
-          impl_get_future() = hpx::make_ready_future<void>();
-#endif
-        });
+  static bool in_parallel(HPX const & = HPX()) noexcept {
+    // TODO: Very awkward to keep track of. What should this really return?
```

Comment:
I've opened https://github.com/kokkos/kokkos/issues/5744 to separately take care of this.